js/logic.js

```diff
@@ -0,0 +1,862 @@
+    this.maxDepth = opts.maxDepth || 10;
+    this.maxStringLength = opts.maxStringLength || 1000;
+    this.maxArrayLength = opts.maxArrayLength || 1000;
+    this.maxObjectLength = opts.maxObjectLength || 10;
+  }
+
+  ObjectSimplifier.prototype = {
+    simplify: function(x) {
+      return this._simplify(x, 0, new WeakSet());
+    },
+
+    _simplify: function(x, depth, cacheSet) {
+      if (cacheSet.has(x)) {
+        return '(cycle)';
+      }
+      if (typeof x === 'number') {
+        return x;
+      } else if (typeof x === 'string') {
+        return x.slice(0, this.maxStringLength);
+      } else if (x && x.BYTES_PER_ELEMENT) {
+        // TypedArray
+        return x.slice(0, this.maxArrayLength);
+      } else if (Array.isArray(x)) {
+        if (depth < this.maxDepth) {
+          return x.slice(0, this.maxArrayLength)
+            .map((element) => this._simplify(element, depth + 1, cacheSet));
+        } else {
+          return '[Array length=' + x.length + ']';
+        }
+      } else if (x && typeof x === 'object') {
+        cacheSet.add(x);
+        if (!isPlainObject(x)) {
+          if (x.toJSON) {
+            return this._simplify(x.toJSON(), depth, cacheSet);
+          } else if (x.toString) {
+            return this._simplify(x.toString(), depth, cacheSet);
+          } else {
+            return '(?)';
+          }
+        } else {
+          if (depth < this.maxDepth) {
+            var retObj = {};
+            var idx = 0;
+            for (var key in x) {
+              if (idx > this.maxObjectLength) {
+                break;
+              }
+              retObj[key] = this._simplify(x[key], depth + 1, cacheSet);
+              idx++;
+            }
+            return retObj;
+          } else if (x.toString) {
+            return this._simplify(x.toString(), depth, cacheSet);
+          } else {
+            return '(object?)';
+          }
+        }
+      } else if (typeof x === 'function') {
+        return '(function)';
+      } else {
+        return x;
+      }
+    }
+  }
+
+  function LogicEvent(scope, type, details) {
+    if (!(scope instanceof Scope)) {
+      throw new Error('Invalid "scope" passed to LogicEvent(); ' +
+                      'did you remember to call logic.defineScope()?');
+    }
+
+    this.scope = scope;
+    this.type = type;
+    this.details = details;
+    this.time = Date.now();
+    this.id = logic.uniqueId();
+  }
+
+  LogicEvent.fromJSON = function(data) {
+    var event = new LogicEvent(new Scope(data.namespace),
+                               data.type,
+                               data.details);
+    event.time = data.time;
+    event.id = data.id;
+    return event;
+  }
+
+  LogicEvent.prototype = {
+    get namespace() {
+      return this.scope.namespace;
+    },
+
+    toJSON: function() {
+      var event = {
+        namespace: this.scope.namespace,
+        type: this.type,
+        details: new ObjectSimplifier().simplify(this.details),
```

Comment:
Mix-up complication: My brain was thinking "type" in logic.event() is the namespace.  That's clearly not the case.  But for our logging needs, I think it might also make sense to emit based on the namespace.  We can of course do complicated things about pre-computing log levels and such based on scopes, but I would postpone all that for future-work.  The "does the event emitter care?" heuristic still seems like a reasonable approach in terms of simplicity and foolproofness.